### ISSUE TITLE
Can't map one shader as a sampler2D texture buffer onto another

### ISSUE THREAD

**Issue by Tremeschin** [polyfloyd/shady#11]
First of all, 

- [x] I have read on the README.md that `NOTE: Buffer support is not very well tested, your mileage may vary.`

_.. however_ I really need this (sorta).

### Context

I'm coding a sorta "video editor" / layers compositor for music visualization and having layers is essential to this task, not only I can separate the elements but apply effects on each and individually.

It did work at first by calling (recursively) what ever the previous layer have to offer, wrote some Python that creates many `vec4 layerN` functions and I tie them all together by calling `vec4 previous_layer = layer{N-1}()`.

Though while this worked so far, when I got to code some multi sampling or gaussian blur filters, things got a bit scary because the loops explode in compute count very quickly due to the nature of this recursive calls.

In the end my GPU could only give me 6 frames per second of a single background image with gaussblur (kernel size 5x5) and a logo at the middle, multisampling the surrounding pixels radially with 20 points or so because rotation aliasing.

That was when I got to search if I could map a function into some `sampler2D` object (spoiler: I couldn't). Then I go to RTFM yours README.md and found about the buffer mapping. This would speed up stuff drastically since I won't be recomputing lots of points.

### Issue / reproducing

I tried manually to set up two shaders with almost the same code called `layer1.glsl` and `layer2.glsl`:

```glsl
void mainImage(out vec4 fragColor, in vec2 fragCoord) {
    vec2 uv = fragCoord.xy / iResolution.xy;
    fragColor = vec4(0.0, 0.0, uv.y, 1);
}
```

The only difference was that `layer2.glsl` returns ` fragColor = vec4(uv.x, 0.0, 0.0, 1);` so I could mix the two with some `textureLod` method and expect a purple at top right corder, black bottom left, red bottom right and blue top left by adding each other's `vec4 pixel_color`.

However either by defining on the `layer2.glsl` file the map buffer with:

`#pragma map layer1=buffer:/path/to/layer1.glsl;1280x720`

Or by giving to the CLI: `./shady -i layer2.glsl -g 1280x720 -map "layer1=buffer:/path/to/layer1.glsl;1280x720"`

It just didn't work, outputting the following error:

`Error reloading environment: failed to call eglCreateContext`

### What I tried

The error is outputted on the file `/shady/egl/egl.go` line 114

Also if I run it with `EGL_PLATFORM=drm` it says failed to create `eglCreatePbufferSurface`, which is on line 91 of the same file.

I don't know where to even start debugging the code as I can also barely create a functional OpenGL context / surface with my main language Python.

Also tried some others GLSL versions by adding `-glsl (version)`, max working version is `410` otherwise it just yells no function / variables called

### What solves this issue:

- Convert `vec4 function() -> sampler2D` on the GLSL code itself (dunno if doable)

- Yours buffer map working properly

I suspect some sorta width or height is being passed wrongly (if you create a separate GL context for rasterizing the buffer map shader at all).

Don't know [how much / when] you tested this feature or if you even care on having it, personally I would appreciate a lot it working since it seems to fix 80% of my current problems, the other 20% are learning shaders and keeping stuff organized since this performance hit is about all what I care for even trying shaders in the first place, also much more flexibility and wayy less code to do stuff.

Thanks!!

**Comment by Tremeschin** [comment]
Got some progress on this, I added a few `fmt.Printf` into that `egl.go CreateContext` function and noticed that the major and minor version were a bit random with some high values integers, probably some pointer to a variable that wasn't properly initialized?

Either way on the file `/shady/renderer/renderer.go` on the function `initEGL` on line 59 I added `glVersion = OpenGL33`, while it does load the context and executes the base input shader properly and I get the right OpenGL major / minor version on `CreateContext` it doesn't seem to be accessing the other's buffer shader content. Will report back any extra progress.

**Comment by Tremeschin** [comment]
Got some progress on this, the OnScreenEngine glVersion was zero [in this line](https://github.com/polyfloyd/shady/blob/ecc25306726fe627e6e04adef70977e678db7ead/renderer/renderer.go#L563), looks like you forgot to initialize the engine's glVersion variable [here](https://github.com/polyfloyd/shady/blob/ecc25306726fe627e6e04adef70977e678db7ead/renderer/renderer.go#L388) :p, adding `glVersion: glVersion,` to the struct initialization fixes this gl versioning and creates the shader with the right glVersion, however getting pixels out of this back buffer is all black (not being rasterized?).. will report back eny extra progress

**Comment by polyfloyd** [comment]
That TODO is for the OnScreenEngine, if you render to stdout or files you will not encounter that.

**Comment by Tremeschin** [comment]
> That TODO is for the OnScreenEngine, if you render to stdout or files you will not encounter that.

Oh I see, I was very distracted by getting instant feedback on code I was only rendering to files for showing a few friends some progress, didn't cross my mind trying that or peeking at the implementation for the non- on screen render methods. I think I'll use FFplay then to also have this instant feedback on the shaders for the time being.

Will edit the previous comments to make it more concise also. Any plans on implementing it for OnScreenEngine? I'm not in a hurry as my target are rendered videos anyways :p

**Comment by polyfloyd** [comment]
No, I have no plans to implement it. I would like to work on some new stuff so Shady is going to take a backseat
